src/libutil/input-accessor.hh

```diff
@@ -0,0 +1,27 @@
+#pragma once
+///@file
+
+#include "source-accessor.hh"
+#include "ref.hh"
+#include "repair-flag.hh"
+
+namespace nix {
+
+MakeError(RestrictedPathError, Error);
+
+struct InputAccessor : virtual SourceAccessor, std::enable_shared_from_this<InputAccessor>
```

Comment:
All the virtual diamond inheritence this has been requiring is a bit cumbersome.

As far as I can tell:

 - `fingerprint` can just go in `SourceAccessor`

 - `getLastModified` is actually dead code --- and I don't think we should need to add it back but rather it can stay part of `Input` and friends in `libfetchers` (where the other `getLastModified` is today

 - `std::enable_shared_from_this` is only needed by `libexpr`, which wishes to store a raw pointer to one of this in `Value`.

I therefore wonder, can we make something like
```c++
template<typename SA>
struct InputAccessor : std::enable_shared_from_this<InputAccessor>
{
    SourceAccessor & ref; // always set to other field
    SA sourceAccessor;
};
```
in `libexpr` and very carefully downcast it + grab reference? What I litterally wrote is UB, but grrr there has to be a way to make this safe.

It is of course quite easy with more indirection:

```c++
// No problematic `template` needed
struct InputAccessor : std::enable_shared_from_this<InputAccessor>
{
    std::unique_ptr<SourceAccessor> sourceAccessor;
};
```

I guess the most obviously safe way is to do

```c++
struct InputAccessor : std::enable_shared_from_this<InputAccessor>
{
    virtual SourceAccessor & getSourceAccessor() = 0;
};

template<typename SA>
struct InputAccessorImpl : InputAccessor
{
    SA impl;
    SourceAccessor & getSourceAccessor() override
    {
        return impl;
    }
};
```

That's kind of manual virtual inheritance :sweat_smile: but it might still be an improvement. Curious what others think.

Reply:
I'd prefer for this PR to avoid modifying unrelated code as much as possible.

Reply:
Sure